### ISSUE TITLE
test_single_triangle_camera_fisheye.py outputs "odd" result after 05/13/2019 update.

### ISSUE THREAD

**Issue by SuperShinyEyes** [BachiLi/redner#36]
02/02/2019 version and 05/13/2019 version output different test result for `test_single_triangle_camera_fisheye.py`.

Here's the video (left: old, right: new): https://www.dropbox.com/s/f7rwn3m7w5q2pbn/test_single_triangle_camera_fisheye.mp4?dl=0

The old version converges close to target:
![target](https://user-images.githubusercontent.com/5738363/57717189-5166dd80-7683-11e9-85c6-536301a2a1f6.png)
final_diff:
![final_diff](https://user-images.githubusercontent.com/5738363/57717209-5d529f80-7683-11e9-88a9-073b5afbc802.png)

The new version does not converge and the final_diff is very large:
![final_diff](https://user-images.githubusercontent.com/5738363/57717248-752a2380-7683-11e9-867c-8b69b6311358.png)

Reproduce:
```bash
cd tests/
python test_single_triangle_camera_fisheye.py
```

I'm using CPU on both cases.

**Comment by SuperShinyEyes** [comment]
There is a same issue with `test_single_triangle_camera.py`: https://www.dropbox.com/s/196wuhm30sa1hc0/test_single_triangle_camera.mp4?dl=0

Target:
![target](https://user-images.githubusercontent.com/5738363/57718140-54fb6400-7685-11e9-973f-f464bd6d5b35.png)

Old final_diff:
![final_diff](https://user-images.githubusercontent.com/5738363/57718164-5fb5f900-7685-11e9-8979-89c63ae40d70.png)

New final_diff:
![final_diff](https://user-images.githubusercontent.com/5738363/57718203-6e041500-7685-11e9-8d2b-2d62ded04243.png)

**Comment by BachiLi** [comment]
Crap. Let me take a look.

**Comment by BachiLi** [comment]
This is quite interesting. Previously the gradients were "wrong" (i.e. not unbiased) because we propagate the gradients through the look at transform in pytorch. Redner outputs unbiased gradients G for the camera to world martrix, but the look at transform f is non-linear, so E[f(G)] != f(E[G]). I fixed this in the 05/13/2019 update by propagating the gradients w.r.t. look at transform inside redner, and this does output correct gradients (I checked with finite difference).

Unfortunately, having correct gradients doesn't always help optimization. In this case, the parallax effect caused by the camera origin in the look at transform creates very nasty optimization landscape. Somehow the previous incorrect gradients smooth out the loss function in this particular case and make the optimization much more well-behaved.

The key cause to this is the non-linear coupling between the camera origin and the camera matrix in look at transform: the camera origin affects *all* entries in the matrix non-linearly (e.g. your looking direction also changes if you move camera origin). I think the proper fix is to simply not use the traditional look at transform and take camera origin, direction, and up vector as inputs. I'll check this.

**Comment by BachiLi** [comment]
In fact simply adjusting the learning rate fixed the problem. The newest commit should solve this.

**Comment by SuperShinyEyes** [comment]
Did it really fix? Have you checked your `final_diff.png`?
Mine looks like 
![final_diff](https://user-images.githubusercontent.com/5738363/58165125-c17ef000-7c8f-11e9-85ce-9c3947e6e1be.png)

**Comment by BachiLi** [comment]
Yup. Mine looks like
![final_diff](https://user-images.githubusercontent.com/951021/58172428-9338ea80-7c66-11e9-8d0e-cd995953cab8.png)

**Comment by BachiLi** [comment]
Yup. Mine looks like
![final_diff](https://user-images.githubusercontent.com/951021/58172497-c8ddd380-7c66-11e9-8e14-61ecb03f4ef5.png)

Did you pull? git checkout tests/test_single_triangle_camera_fisheye.py ?

**Comment by SuperShinyEyes** [comment]
I manually changed the learning rate since that is supposedly the only update according to your commit. Do I need to change c++ code for this?

**Comment by BachiLi** [comment]
I don't remember. Maybe clone and try to find the difference?

**Comment by SuperShinyEyes** [comment]
Cloned version works fine now. Thanks!
